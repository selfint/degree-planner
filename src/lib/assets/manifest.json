{"2023_2024":{"computer_science":{"3_year":{"recommended":{"semester1":null,"semester2":null,"semester3":null,"semester4":null,"semester5":null,"semester6":null,"semester7":null},"requirements":{"core":{"courses":null,"points":null},"english":{"count":null,"courses":null},"list_a":{"courses":null,"overflow":null,"points":null},"list_b":{"courses":null,"points":null},"malag":{"courses":null,"points":null},"math":{"count":null,"courses":null},"other":{"courses":null,"points":null},"points":null,"project":{"count":null,"courses":null},"science":{"choice":{"amount":null,"biology_chain":{"count":null,"courses":null},"physics-chemistry_chain":{"count":null,"courses":null},"physics_chain":{"choice":{"1":{"count":null,"courses":null},"2":{"count":null,"courses":null},"amount":null}}},"courses":null,"overflow":null,"points":null}}}}}}