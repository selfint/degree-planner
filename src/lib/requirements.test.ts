import fs from 'fs/promises';
import path from 'path';

import { describe, it, vi, beforeEach, afterEach } from 'vitest';

import manifest from '$lib/assets/manifest.json';
import {
	loadDegreeRequirements,
	loadDegreeRecommendation
} from './requirements';

describe('Requirements parser', () => {
	beforeEach(() => {
		// @ts-expect-error
		global.fetch = vi.fn(async (url: string) => {
			const base = path.join(process.cwd(), 'static');
			const filepath = path.join(base, url);

			return {
				ok: true,
				text: async () => fs.readFile(filepath).then((buf) => buf.toString())
			};
		});
	});

	afterEach(() => {
		vi.restoreAllMocks();
	});

	it('should parse recommended', async (ctx) => {
		const degree: Degree = ['2023_2024', 'computer_science', '3_year'];
		const header: Record<string, null> =
			// @ts-expect-error
			manifest[degree[0]][degree[1]][degree[2]].recommended;

		const requirements = await loadDegreeRecommendation(degree, header);

		ctx.expect(requirements).toMatchSnapshot();
	});

	it('should parse requirements', async (ctx) => {
		const degree: Degree = ['2023_2024', 'computer_science', '3_year'];
		const header: RequirementsHeader =
			// @ts-expect-error
			manifest[degree[0]][degree[1]][degree[2]].requirements;

		const requirements = await loadDegreeRequirements(degree, header);

		ctx.expect(requirements).toMatchSnapshot();
	});
});
