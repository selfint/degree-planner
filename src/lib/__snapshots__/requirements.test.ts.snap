// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Requirements parser > should parse recommended 1`] = `
[
  [
    "104031",
    "104166",
    "234114",
    "234129",
    "324033",
  ],
  [
    "104032",
    "114071",
    "234124",
    "234125",
    "234141",
  ],
  [
    "094412",
    "104134",
    "234218",
    "044252",
    "234252",
    "234292",
  ],
  [
    "234118",
    "234123",
    "234247",
  ],
  [
    "236343",
    "236360",
  ],
]
`;

exports[`Requirements parser > should parse requirements 1`] = `
{
  "core": {
    "courses": [
      "104031",
      "104166",
      "234114",
      "234129",
      "324033",
      "134058",
      "134020",
      "104032",
      "114071",
      "234124",
      "234125",
      "234141",
      "094412",
      "104134",
      "234218",
      "044252",
      "234252",
      "234292",
      "104135",
      "104033",
      "104174",
      "104122",
      "104142",
      "104285",
      "104295",
      "236267",
      "236343",
      "236360",
      "114075",
      "114052",
      "114054",
      "114073",
      "114101",
      "114246",
      "124120",
      "125001",
      "125801",
      "124510",
    ],
    "points": 84.5,
  },
  "english": {
    "count": 2,
    "courses": [
      "086762",
      "084515",
      "086320",
      "086478",
      "085915",
      "086800",
      "086822",
      "274251",
      "198000",
      "236009",
      "016102",
      "016208",
      "016328",
      "017012",
      "014603",
      "014143",
      "016503",
      "315242",
      "315014",
      "324262",
      "324282",
      "324305",
      "324527",
      "325016",
      "326001",
      "326002",
      "326006",
      "326009",
      "104122",
      "104181",
      "106350",
      "106395",
      "106429",
      "106803",
      "106935",
      "106937",
      "196012",
      "196014",
      "197007",
      "197014",
      "046200",
      "046240",
      "046746",
      "134039",
      "134128",
      "136037",
      "114229",
      "114252",
      "114250",
      "114251",
      "056149",
      "056391",
      "056404",
      "056406",
      "056407",
      "056403",
      "056405",
      "056146",
      "336014",
      "336032",
      "336209",
      "336539",
      "336541",
      "336549",
      "338001",
      "236025",
      "236306",
      "236613",
      "236640",
      "236664",
      "236824",
      "019627",
      "205586",
      "205028",
      "206926",
      "207048",
      "206006",
      "206840",
      "034034",
      "036001",
      "036004",
      "036008",
      "036010",
      "036013",
      "036014",
      "036026",
      "036027",
      "036032",
      "036035",
      "036044",
      "036063",
      "036065",
      "036079",
      "036080",
      "036711",
      "036097",
      "036103",
      "066605",
      "066532",
      "066533",
      "126200",
      "126603",
      "126700",
      "127002",
      "127446",
      "127727",
      "096226",
      "096412",
      "096606",
    ],
  },
  "list_a": {
    "courses": [
      "234268",
      "234302",
      "234303",
      "234304",
      "234313",
      "234326",
      "234329",
      "234493",
      "234901",
      "236002",
      "236004",
      "236025",
      "236026",
      "236201",
      "236203",
      "236204",
      "236205",
      "236206",
      "236216",
      "236268",
      "236270",
      "236271",
      "236272",
      "236278",
      "236299",
      "236303",
      "236304",
      "236306",
      "236309",
      "236310",
      "236313",
      "236315",
      "236318",
      "236319",
      "236321",
      "236322",
      "236323",
      "236324",
      "236328",
      "236329",
      "236330",
      "236332",
      "236333",
      "236334",
      "236336",
      "236340",
      "236341",
      "236342",
      "236345",
      "236346",
      "236347",
      "236348",
      "236349",
      "236350",
      "236351",
      "236356",
      "236357",
      "236358",
      "236359",
      "236360",
      "236361",
      "236363",
      "236366",
      "236369",
      "236370",
      "236371",
      "236372",
      "236373",
      "236374",
      "236376",
      "236377",
      "236378",
      "236379",
      "236381",
      "236388",
      "236422",
      "236490",
      "236491",
      "236496",
      "236499",
      "236500",
      "236501",
      "236502",
      "236503",
      "236504",
      "236506",
      "236508",
      "236509",
      "236510",
      "236512",
      "236513",
      "236515",
      "236518",
      "236520",
      "236521",
      "236522",
      "236523",
      "236524",
      "236525",
      "236526",
      "236612",
      "236613",
      "236620",
      "236621",
      "236622",
      "236623",
      "236624",
      "236625",
      "236627",
      "236628",
      "236629",
      "236630",
      "236631",
      "236632",
      "236633",
      "236634",
      "236635",
      "236637",
      "236638",
      "236640",
      "236641",
      "236643",
      "236644",
      "236645",
      "236646",
      "236647",
      "236648",
      "236649",
      "236650",
      "236651",
      "236652",
      "236653",
      "236654",
      "236655",
      "236657",
      "236658",
      "236660",
      "236661",
      "236662",
      "236663",
      "236664",
      "236667",
      "236668",
      "236669",
      "236670",
      "236698",
      "236700",
      "236703",
      "236712",
      "236715",
      "236716",
      "236719",
      "236729",
      "236754",
      "236755",
      "236756",
      "236757",
      "236760",
      "236777",
      "236779",
      "236780",
      "236781",
      "236800",
      "236811",
      "236812",
      "236813",
      "236814",
      "236815",
      "236816",
      "236817",
      "236818",
      "236819",
      "236820",
      "236821",
      "236822",
      "236823",
      "236824",
      "236825",
      "236826",
      "236827",
      "236828",
      "236829",
      "236830",
      "236831",
      "236832",
      "236833",
      "236834",
      "236835",
      "236836",
      "236837",
      "236838",
      "236860",
      "236861",
      "236862",
      "236873",
      "236874",
      "236875",
      "236901",
      "236927",
      "236990",
      "236991",
      "238100",
      "238125",
      "238739",
      "238790",
      "238900",
      "238901",
      "238902",
    ],
    "overflow": "list_b",
    "points": 18,
  },
  "list_b": {
    "courses": [
      "036044",
      "044105",
      "044127",
      "044131",
      "044137",
      "044157",
      "044167",
      "044169",
      "044202",
      "046201",
      "046206",
      "046332",
      "046880",
      "048878",
      "048921",
      "086761",
      "094222",
      "094313",
      "094314",
      "094333",
      "094334",
      "094423",
      "094591",
      "096200",
      "096211",
      "096224",
      "096250",
      "096262",
      "096326",
      "096411",
      "097317",
      "104122",
      "104135",
      "104142",
      "104157",
      "104165",
      "104174",
      "104158",
      "104177",
      "104192",
      "104221",
      "104223",
      "104279",
      "104293",
      "106378",
      "104294",
      "106383",
      "114101",
      "114246",
      "115203",
      "115204",
      "114036",
      "116217",
      "116354",
      "124120",
      "124400",
      "124503",
      "124801",
      "125801",
      "134019",
      "134020",
      "134058",
      "134082",
      "134113",
      "134128",
      "134119",
      "134142",
      "214909",
      "104033",
      "104285",
      "104295",
    ],
    "points": 6.5,
  },
  "malag": {
    "courses": [
      "214120",
      "324258",
      "324262",
      "324265",
      "324269",
      "324273",
      "324282",
      "324284",
      "324286",
      "324292",
      "324298",
      "324305",
      "324306",
      "324307",
      "324315",
      "324436",
      "324441",
      "324442",
      "324445",
      "324474",
      "324518",
      "324520",
      "324527",
      "324528",
      "324540",
      "324541",
      "324962",
      "324987",
      "324992",
      "325001",
      "325002",
      "325006",
      "325010",
      "325011",
      "325012",
      "325015",
      "325016",
      "325017",
      "325019",
      "325020",
      "326001",
      "326002",
      "326006",
      "326009",
    ],
    "points": 6,
  },
  "math": {
    "count": 1,
    "courses": [
      "104135",
      "104033",
      "104174",
      "104122",
      "104142",
      "104285",
      "104295",
    ],
  },
  "other": {
    "courses": [],
    "points": 2,
  },
  "points": 118.5,
  "project": {
    "count": 1,
    "courses": [
      "234302",
      "234303",
      "234304",
      "234313",
      "234326",
      "234329",
      "236270",
      "236272",
      "236303",
      "236323",
      "236328",
      "236333",
      "236340",
      "236346",
      "236349",
      "236361",
      "236366",
      "236371",
      "236381",
      "236499",
      "236502",
      "236503",
      "236504",
      "236512",
      "236513",
      "236524",
      "236526",
      "236729",
      "236754",
      "236757",
      "236828",
      "236874",
      "236991",
      "044167",
      "044169",
      "114252",
    ],
  },
  "science": {
    "choice": {
      "amount": 1,
      "biology_chain": {
        "count": 1,
        "courses": [
          "134058",
          "134020",
        ],
      },
      "physics-chemistry_chain": {
        "count": 1,
        "courses": [
          "124120",
          "114052",
        ],
      },
      "physics_chain": {
        "choice": {
          "1": {
            "count": 1,
            "courses": [
              "114075",
            ],
          },
          "2": {
            "count": 2,
            "courses": [
              "114052",
              "114054",
            ],
          },
          "amount": 1,
        },
      },
    },
    "courses": [
      "114075",
      "114052",
      "114054",
      "114073",
      "114101",
      "114246",
      "124120",
      "125001",
      "125801",
      "124510",
      "134058",
      "134020",
    ],
    "overflow": "list_b",
    "points": 8,
  },
}
`;
